wsmux/session.go

```diff
@@ -0,0 +1,237 @@
+package wsmux
+
+import (
+	"encoding/binary"
+	// "io"
+	"io/ioutil"
+	// "log"
+	"net"
+	// "runtime"
+	"bytes"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	// "github.com/taskcluster/slugid-go/slugid"
+)
+
+const (
+	defaultQueueSize       = 10
+	defaultStreamQueueSize = 10
+)
+
+type Session struct {
+	mu       sync.Mutex
+	streams  map[uint32]*stream
+	streamCh chan *stream
+	writes   chan frame
+	conn     *websocket.Conn
+
+	acceptDeadline time.Time
+	readDeadline   time.Time
+	writeDeadline  time.Time
+
+	logger Logger
+
+	closed       atomic.Value
+	remoteClosed atomic.Value
+
+	nextID uint32
+
+	remoteCloseCallback func()
+}
+
+func newSession(conn *websocket.Conn, server bool, conf *Config) *Session {
+	if conf == nil {
+		conf = &Config{}
+	}
+	s := &Session{}
+	s.conn = conn
+	s.streams = make(map[uint32]*stream)
+	s.streamCh = make(chan *stream, defaultStreamQueueSize)
+	s.writes = make(chan frame, defaultQueueSize)
+
+	s.acceptDeadline = conf.AcceptDeadline
+	s.readDeadline = conf.ReadDeadline
+	s.writeDeadline = conf.WriteDeadline
+	s.remoteCloseCallback = conf.RemoteCloseCallback
+
+	s.closed.Store(false)
+	s.remoteClosed.Store(false)
+
+	if server {
+		s.nextID = 0
+	} else {
+		s.nextID = 1
+	}
+
+	if conf.Log == nil {
+		s.logger = &nilLogger{}
+	} else {
+		s.logger = conf.Log
+	}
+	go s.sendLoop()
+	go s.recvLoop()
+	return s
+}
+
+func (s *Session) Accept() (net.Conn, error) {
+	var timeout <-chan time.Time
+	var timer *time.Timer
+	if !s.acceptDeadline.IsZero() {
+		timer = time.NewTimer(s.acceptDeadline.Sub(time.Now()))
+		timeout = timer.C
+	}
+
+	select {
+	case <-timeout:
+		return nil, ErrAcceptTimeout
+	case str := <-s.streamCh:
+		return str, nil
+	}
+}
+
+func (s *Session) Open() (net.Conn, error) {
+	id := atomic.LoadUint32(&s.nextID)
```

Comment:
Don't use an atomic...
Just because there is no race condition as far as the race detector is aware you still have a race condition :)

Two concurrent calls to Open() can get the same ID.
This is why you rarely need atomics... You are much better of saying that `mu.Lock()` must be called before accessing `nextID`.

Also please use some empty lines to separeret blocks... You have 3 blocks here:
 1. Finding the ID
 2. Sending first message
 3. Creating stream and adding it to internal mapping.

Each of these should start with an empty line and a one line comment... Otherwise it becomes hard to read.

However, you probably have additional race conditions because of the 3 block split. You probably want to create the stream object, insert it in internal mapping and then send first message.

Finally, you likely have a deadlock, if `s.writes` is closed then what happens ? What happens if the underlying connection is closed?
You probably want to do a select statements over `s.writes` and some other gochannel on the `Session` object which is used to communicate connection closure..